sockeye/inference.py

```diff
@@ -709,26 +785,30 @@ def _concat_translations(self, translations: List[Translation]) -> Translation:
 
     def translate_nd(self,
                      source: mx.nd.NDArray,
-                     source_length: int) -> Translation:
+                     source_length: int) -> List[Translation]:
         """
         Translates source of source_length, given a bucket_key.
 
-        :param source: Source ids. Shape: (1, bucket_key).
+        :param source: Source ids. Shape: (batch_size, bucket_key) or (1, bucket_key)
         :param source_length: Bucket key.
 
         :return: Sequence of translated ids, attention matrix, length-normalized negative log probability.
         """
         return self._get_best_from_beam(*self._beam_search(source, source_length))
 
-    def _encode(self, source: mx.nd.NDArray, source_length: int) -> List[ModelState]:
+    def _encode(self, sources: mx.nd.NDArray, source_length: int) -> List[ModelState]:
         """
         Returns a ModelState for each model representing the state of the model after encoding the source.
 
-        :param source: Source ids. Shape: (1, bucket_key).
+        :param sources: Source ids. Shape: (num_sent, bucket_key).
         :param source_length: Bucket key.
         :return: List of ModelStates.
         """
-        return [ModelState(states=m.run_encoder(source, source_length)) for m in self.models]
+        states = [ModelState()] * len(self.models)
+        for model, state in zip(self.models, states):
+            for src in sources:
+                state.append_state(model.run_encoder(mx.nd.expand_dims(src, axis=0), source_length))
```

Comment:
I don't like this iterative invocation of the encoder module. 
This really should be batched as well.

Reply:
Ok, will rewrite this